### PR TITLE
[CMake] Enable NCCL only when enabling CUDA or ROCm support

### PR DESCRIPTION
**Pull request by alberto-magni**
Conditionally enable NCCL depending on CUDA and ROCM

Before this change NCCL support was enabled unconditionally, even
when building without CUDA or ROCM support.
This caused the command:
`$ ./build.sh --enable_training`

To trigger the following cmake warning
-- Could NOT find NCCL (missing: NCCL_INCLUDE_DIR NCCL_LIBRARY)
CMake Warning at CMakeLists.txt:1282 (message):
  NCCL is not found.  Please use --nccl_home to specify the path of NCCL.
  Otherwise, NCCL is disabled.

This is a spurious warning because the user did not ask to search for NCCL.

### CODE REVIEW

**Comment by alberto-magni** [comment]
An alternative solution would be to explicitly require to pass --enable_nccl to build.sh when building with cuda.
(This would require to change the build scripts though)
I just find it odd that we should explicitly pass --disable_nccl when building without cuda support.

**Comment by KeDengMS** [comment]
@snnn could you take a look at this build failure in orttraining-linux-ci-pipeline? Seems not related to this PR.

Unable to find image 'onnxruntimeregistry.azurecr.io/internal/azureml/onnxruntimecpubuild:ch5e' locally

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -1062,6 +1062,9 @@ if (onnxruntime_USE_CUDA)
     list(APPEND ONNXRUNTIME_CUDA_LIBRARIES cublas cudnn curand cufft)
   endif()
 
+  # Enable NCCL by default when using CUDA.
+  set(onnxruntime_USE_NCCL ON)
```
Comment (weixingzhang) [inline cmake/CMakeLists.txt:1066]: NCCL is only used in ORT training now. Could you move this line to line #1195 below? It should be set when MPI is found.

Comment (alberto-magni) [reply cmake/CMakeLists.txt]: Done. When moving this line inside the MPI I enable NCCL only when CUDA is ON.

**Review by weixingzhang — DISMISSED** [review]

**Comment by alberto-magni** [comment]
The linux gpu failed for a timeout: https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=239613&view=logs&j=52024a78-4e9d-503c-68de-2b0746163a14&t=23786350-27d9-5ceb-4348-014ae45926cf
but all the tests are green. Is this common ?

The linux ci pipeline instead fails for a permission issue.

**Comment by snnn** [comment]
> @

Please see the discussion in Teams. Xueyun should already fixed it.

**Comment by alberto-magni** [comment]
Can someone please trigger the build ? I hope the spurious failures are solved now.
Thanks

**Comment by snnn** [comment]
Scheduled.

**Review by weixingzhang — APPROVED** [review]

**Pull request merged**